_posts/2023-08-02-turbocharging-rust-code-verification.md

```diff
@@ -0,0 +1,446 @@
+Similarly, for `sync::spsc::tests::alloc_test`, verification time goes down from 1004 seconds with MiniSat to 63 seconds with Kissat (a 16X speedup), and 70 seconds with CaDiCaL (a 14X speedup).
+
+We also see that MiniSat remains the fastest solver for harnesses that already run in a matter of seconds.
+Of the harnesses that run for more than 10 seconds, Kissat is the fastest on 47% of them, CaDiCaL is the fastest on 24%, and MiniSat on the rest.
+
+By picking the best solver for each harness using the `kani::solver` attribute, we can can bring the total cumulative runtime from 2 hours and 20 minutes down to 15 minutes (counting timeouts as 1800s), while solving two more harnesses. Great savings if your harnesses are run in CI!
+
+[^1]: Without our enhancement to CBMC, it was already possible to select a different SAT solver without rebuilding CBMC via the `--external-sat-solver` option. However, this option doesn't use the solver in incremental mode(i.e. through its library API, keeping the solver alive between successive calls), and instead relies on writing DIMACS files to disk, which often results in decreased performance.
+
+## Adding Direct Export of GOTO Binaries
+
+`kani-compiler` translates a Rust program into a GOTO program and exports it as a *symbol table* for CBMC to analyse. A *symbol table* stores the definitions and source locations of all symbols manipulated by the program (types symbols and their definitions, static symbols and their initial value, functions symbols and their bodies, etc.).
+The symbol table contents hence mostly consists of serialized abstract syntax trees.
+Originally, Kani serialized symbol tables to JSON and then used the CBMC utility `symtab2gb` to translate the JSON symbol table into a *GOTO binary*, which is the format that CBMC actually expects as input.
+The JSON to GOTO binary conversion was one of the most time consuming steps in the compilation stage.
+We implemented direct GOTO binary serialisation in Kani, which allows us to skip the costly invocation of `symtab2gb`.
+Kani can now perform the MIR-to-GOTO code generation and GOTO binary export 4x faster than before.
+
+The table below reports total time and memory consumption when running `kani --tests --only-codegen` with `kani-0.33.3` for three crates of the `s2n-quic` project, with JSON symbol table export versus GOTO binary export.
+
+| Crate                | User time (JSON) | User time (GOTO bin.) | Peak Mem. (JSON) | Peak Mem. (GOTO bin.) |
+| -------------------  | ---------------: | --------------------: | ---------------: | --------------------: |
+| `s2n-quic-core`      |             198s |              **111s** |        **669Mb** |                 815Mb |
+| `s2n-quic-platform`  |              99s |               **91s** |        **449Mb** |                 450Mb |
+| `s2n-quic-xdp`       |              92s |               **95s** |        **450Mb** |                 450Mb |
```

Comment:
Peak memory is being reported in Mb, but aren't the measures reported with the `time` command in MB?

Reply:
fixed